notebooks/customizing_potentials_cookbook.ipynb

```diff
@@ -0,0 +1,1780 @@
+        "colab": {}
+      },
+      "source": [
+        "energy_fn = harmonic_morse_pair(displacement, species=species, D0=5.0, \n",
+        "                                alpha=10.0, r0=r0_species_matrix, k=500.0)\n",
+        "\n",
+        "Rfinal, max_force_component = run_minimization(energy_fn, R, shift)\n",
+        "print('largest component of force after minimization = {}'.format(max_force_component))\n",
+        "\n",
+        "plot_system( Rfinal, box_size, species=species )"
+      ],
+      "execution_count": 0,
+      "outputs": []
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "vvjaCB0cN8rA",
+        "colab_type": "text"
+      },
+      "source": [
+        "##Dynamic Species"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "jyYqW7dCDi0y",
+        "colab_type": "text"
+      },
+      "source": [
+        "Just like standard parameters, the species list can be passed dynamically as well. However, unlike standard parameters, you have to tell ```smap.pair``` that the species will be specified dynamically. To do this, set  ```species=quantity.Dynamic``` when creating your energy function.\n",
+        "\n",
+        "The following sets up an energy function where the attractive part of the interaction only exists between members of the first species, but where the species will be defined dynamically."
+      ]
+    },
+    {
+      "cell_type": "code",
+      "metadata": {
+        "id": "Q30CEsxaQGc7",
+        "colab_type": "code",
+        "colab": {}
+      },
+      "source": [
+        "D0_species_matrix = np.array([[ 5.0, 0.0],\n",
+        "                              [0.0,  0.0]])\n",
+        "\n",
+        "energy_fn = harmonic_morse_pair(displacement, species=quantity.Dynamic, \n",
+        "                                D0=D0_species_matrix, alpha=10.0, r0=0.5, k=500.0)"
+      ],
+      "execution_count": 0,
+      "outputs": []
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "a2_Er0ttWcmu",
+        "colab_type": "text"
+      },
+      "source": [
+        "Now we set up a finite temperature Brownian Dynamics simulation where, at every step, particles on the left half of the simulation box are assigned to species 0, while particles on the right half are assigned to species 1."
+      ]
+    },
+    {
+      "cell_type": "code",
+      "metadata": {
+        "id": "HlT0lZ0IO9rx",
+        "colab_type": "code",
+        "colab": {}
+      },
+      "source": [
+        "def run_brownian(energy_fn, R_init, shift, key, num_steps):\n",
+        "  init, apply = simulate.brownian(energy_fn, shift, dt=0.00001, T_schedule=1.0, gamma=0.1)\n",
+        "  #apply = jit(apply)\n",
+        "\n",
+        "  #Define a function to recalculate the species each step\n",
+        "  def Get_species(R):\n",
+        "    return np.where(R[:,0]<box_size/2,0,1)\n",
+        "\n",
+        "  #@jit\n",
```

Comment:
I don't know if they are superfluous or not, but they do _something_ because when I include them I get the following error message:

```

NOT RESOLVED.
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-39-8c573c874c50> in <module>()
      1 key, split = random.split(key)
----> 2 Rfinal, max_force_component = run_brownian(energy_fn, R, shift, split, num_steps=10000)
      3 plot_system( Rfinal, box_size )

35 frames
<ipython-input-38-eba8c16ddacb> in run_brownian(energy_fn, R_init, shift, key, num_steps)
     17   state = init(split, R_init)
     18 
---> 19   state, _ = lax.scan(scan_fn,state,np.arange(num_steps))
     20   return state.position,np.amax(np.abs(-grad(energy_fn)(state.position,
     21                                                         species=get_species(state.position),

/usr/local/lib/python3.6/dist-packages/jax/lax/lax_control_flow.py in scan(f, init, xs, length)
    934   x_dtypes = [x.dtype for x in xs_flat]
    935   x_avals = tuple(_map(ShapedArray, x_shapes, x_dtypes))
--> 936   jaxpr, consts, out_tree = _initial_style_jaxpr(f, in_tree, carry_avals + x_avals)
    937   out_tree_children = out_tree.children()
    938   if len(out_tree_children) != 2:

/usr/local/lib/python3.6/dist-packages/jax/lax/lax_control_flow.py in _initial_style_jaxpr(fun, in_tree, in_avals)
     62   with core.initial_style_staging():
     63     jaxpr, out_pvals, consts = pe.trace_to_jaxpr(
---> 64       wrapped_fun, in_pvals, instantiate=True, stage_out=False)
     65   out_avals = _map(raise_to_shaped, unzip2(out_pvals)[0])
     66   const_avals = tuple(raise_to_shaped(core.get_aval(c)) for c in consts)

/usr/local/lib/python3.6/dist-packages/jax/interpreters/partial_eval.py in trace_to_jaxpr(fun, pvals, instantiate, stage_out, bottom, trace_type)
    435   with new_master(trace_type, bottom=bottom) as master:
    436     fun = trace_to_subjaxpr(fun, master, instantiate)
--> 437     jaxpr, (out_pvals, consts, env) = fun.call_wrapped(pvals)
    438     assert not env
    439     del master

/usr/local/lib/python3.6/dist-packages/jax/linear_util.py in call_wrapped(self, *args, **kwargs)
    148     gen = None
    149 
--> 150     ans = self.f(*args, **dict(self.params, **kwargs))
    151     del args
    152     while stack:

/usr/local/lib/python3.6/dist-packages/jax/api.py in f_jitted(*args, **kwargs)
    152     flat_fun, out_tree = flatten_fun(f, in_tree)
    153     out = xla.xla_call(flat_fun, *args_flat, device=device, backend=backend,
--> 154                        name=flat_fun.__name__)
    155     return tree_unflatten(out_tree(), out)
    156 

/usr/local/lib/python3.6/dist-packages/jax/core.py in _call_bind(processor, post_processor, primitive, f, *args, **params)
   1003     tracers = map(top_trace.full_raise, args)
   1004     process = getattr(top_trace, processor)
-> 1005     outs = map(full_lower, process(primitive, f, tracers, params))
   1006   return apply_todos(env_trace_todo(), outs)
   1007 

/usr/local/lib/python3.6/dist-packages/jax/interpreters/partial_eval.py in process_call(self, call_primitive, f, tracers, params)
    175     in_pvs, in_consts = unzip2([t.pval for t in tracers])
    176     fun, aux = partial_eval(f, self, in_pvs)
--> 177     out_flat = call_primitive.bind(fun, *in_consts, **params)
    178     out_pvs, jaxpr, env = aux()
    179     env_tracers = map(self.full_raise, env)

/usr/local/lib/python3.6/dist-packages/jax/core.py in _call_bind(processor, post_processor, primitive, f, *args, **params)
    999   if top_trace is None:
   1000     with new_sublevel():
-> 1001       outs = primitive.impl(f, *args, **params)
   1002   else:
   1003     tracers = map(top_trace.full_raise, args)

/usr/local/lib/python3.6/dist-packages/jax/interpreters/xla.py in _xla_call_impl(fun, device, backend, name, *args)
    460 
    461 def _xla_call_impl(fun: lu.WrappedFun, *args, device, backend, name):
--> 462   compiled_fun = _xla_callable(fun, device, backend, name, *map(arg_spec, args))
    463   try:
    464     return compiled_fun(*args)

/usr/local/lib/python3.6/dist-packages/jax/linear_util.py in memoized_fun(fun, *args)
    219       fun.populate_stores(stores)
    220     else:
--> 221       ans = call(fun, *args)
    222       cache[key] = (ans, fun.stores)
    223     return ans

/usr/local/lib/python3.6/dist-packages/jax/interpreters/xla.py in _xla_callable(fun, device, backend, name, *arg_specs)
    477   pvals: Sequence[pe.PartialVal] = [pe.PartialVal.unknown(aval) for aval in abstract_args]
    478   jaxpr, pvals, consts = pe.trace_to_jaxpr(
--> 479       fun, pvals, instantiate=False, stage_out=True, bottom=True)
    480 
    481   _map(prefetch, it.chain(consts, jaxpr_literals(jaxpr)))

/usr/local/lib/python3.6/dist-packages/jax/interpreters/partial_eval.py in trace_to_jaxpr(fun, pvals, instantiate, stage_out, bottom, trace_type)
    435   with new_master(trace_type, bottom=bottom) as master:
    436     fun = trace_to_subjaxpr(fun, master, instantiate)
--> 437     jaxpr, (out_pvals, consts, env) = fun.call_wrapped(pvals)
    438     assert not env
    439     del master

/usr/local/lib/python3.6/dist-packages/jax/linear_util.py in call_wrapped(self, *args, **kwargs)
    148     gen = None
    149 
--> 150     ans = self.f(*args, **dict(self.params, **kwargs))
    151     del args
    152     while stack:

<ipython-input-38-eba8c16ddacb> in scan_fn(state, t)
     12     species = get_species(state.position)
     13     #Dynamically pass species to apply, which passes it on to energy_fn
---> 14     return apply(state, species=species, species_count=2), 0
     15 
     16   key, split = random.split(key)

/usr/local/lib/python3.6/dist-packages/jax/api.py in f_jitted(*args, **kwargs)
    152     flat_fun, out_tree = flatten_fun(f, in_tree)
    153     out = xla.xla_call(flat_fun, *args_flat, device=device, backend=backend,
--> 154                        name=flat_fun.__name__)
    155     return tree_unflatten(out_tree(), out)
    156 

/usr/local/lib/python3.6/dist-packages/jax/core.py in _call_bind(processor, post_processor, primitive, f, *args, **params)
   1003     tracers = map(top_trace.full_raise, args)
   1004     process = getattr(top_trace, processor)
-> 1005     outs = map(full_lower, process(primitive, f, tracers, params))
   1006   return apply_todos(env_trace_todo(), outs)
   1007 

/usr/local/lib/python3.6/dist-packages/jax/interpreters/partial_eval.py in process_call(self, call_primitive, f, tracers, params)
    175     in_pvs, in_consts = unzip2([t.pval for t in tracers])
    176     fun, aux = partial_eval(f, self, in_pvs)
--> 177     out_flat = call_primitive.bind(fun, *in_consts, **params)
    178     out_pvs, jaxpr, env = aux()
    179     env_tracers = map(self.full_raise, env)

/usr/local/lib/python3.6/dist-packages/jax/core.py in _call_bind(processor, post_processor, primitive, f, *args, **params)
   1003     tracers = map(top_trace.full_raise, args)
   1004     process = getattr(top_trace, processor)
-> 1005     outs = map(full_lower, process(primitive, f, tracers, params))
   1006   return apply_todos(env_trace_todo(), outs)
   1007 

/usr/local/lib/python3.6/dist-packages/jax/interpreters/partial_eval.py in process_call(self, call_primitive, f, tracers, params)
    175     in_pvs, in_consts = unzip2([t.pval for t in tracers])
    176     fun, aux = partial_eval(f, self, in_pvs)
--> 177     out_flat = call_primitive.bind(fun, *in_consts, **params)
    178     out_pvs, jaxpr, env = aux()
    179     env_tracers = map(self.full_raise, env)

/usr/local/lib/python3.6/dist-packages/jax/core.py in _call_bind(processor, post_processor, primitive, f, *args, **params)
    999   if top_trace is None:
   1000     with new_sublevel():
-> 1001       outs = primitive.impl(f, *args, **params)
   1002   else:
   1003     tracers = map(top_trace.full_raise, args)

/usr/local/lib/python3.6/dist-packages/jax/interpreters/xla.py in _xla_call_impl(fun, device, backend, name, *args)
    460 
    461 def _xla_call_impl(fun: lu.WrappedFun, *args, device, backend, name):
--> 462   compiled_fun = _xla_callable(fun, device, backend, name, *map(arg_spec, args))
    463   try:
    464     return compiled_fun(*args)

/usr/local/lib/python3.6/dist-packages/jax/linear_util.py in memoized_fun(fun, *args)
    219       fun.populate_stores(stores)
    220     else:
--> 221       ans = call(fun, *args)
    222       cache[key] = (ans, fun.stores)
    223     return ans

/usr/local/lib/python3.6/dist-packages/jax/interpreters/xla.py in _xla_callable(fun, device, backend, name, *arg_specs)
    477   pvals: Sequence[pe.PartialVal] = [pe.PartialVal.unknown(aval) for aval in abstract_args]
    478   jaxpr, pvals, consts = pe.trace_to_jaxpr(
--> 479       fun, pvals, instantiate=False, stage_out=True, bottom=True)
    480 
    481   _map(prefetch, it.chain(consts, jaxpr_literals(jaxpr)))

/usr/local/lib/python3.6/dist-packages/jax/interpreters/partial_eval.py in trace_to_jaxpr(fun, pvals, instantiate, stage_out, bottom, trace_type)
    435   with new_master(trace_type, bottom=bottom) as master:
    436     fun = trace_to_subjaxpr(fun, master, instantiate)
--> 437     jaxpr, (out_pvals, consts, env) = fun.call_wrapped(pvals)
    438     assert not env
    439     del master

/usr/local/lib/python3.6/dist-packages/jax/linear_util.py in call_wrapped(self, *args, **kwargs)
    148     gen = None
    149 
--> 150     ans = self.f(*args, **dict(self.params, **kwargs))
    151     del args
    152     while stack:

/usr/local/lib/python3.6/dist-packages/jax_md/simulate.py in apply_fn(state, t, **kwargs)
    457     key, split = random.split(key)
    458 
--> 459     F = force_fn(R, t=t, **kwargs)
    460     xi = random.normal(split, R.shape, R.dtype)
    461 

/usr/local/lib/python3.6/dist-packages/jax/api.py in grad_f(*args, **kwargs)
    379   @wraps(fun, docstr=docstr, argnums=argnums)
    380   def grad_f(*args, **kwargs):
--> 381     _, g = value_and_grad_f(*args, **kwargs)
    382     return g
    383 

/usr/local/lib/python3.6/dist-packages/jax/api.py in value_and_grad_f(*args, **kwargs)
    436     f_partial, dyn_args = argnums_partial(f, argnums, args)
    437     if not has_aux:
--> 438       ans, vjp_py = _vjp(f_partial, *dyn_args)
    439     else:
    440       ans, vjp_py, aux = _vjp(f_partial, *dyn_args, has_aux=True)

/usr/local/lib/python3.6/dist-packages/jax/api.py in _vjp(fun, *primals, **kwargs)
   1437   if not has_aux:
   1438     flat_fun, out_tree = flatten_fun_nokwargs(fun, in_tree)
-> 1439     out_primal, out_vjp = ad.vjp(flat_fun, primals_flat)
   1440     out_tree = out_tree()
   1441   else:

/usr/local/lib/python3.6/dist-packages/jax/interpreters/ad.py in vjp(traceable, primals, has_aux)
    104 def vjp(traceable, primals, has_aux=False):
    105   if not has_aux:
--> 106     out_primals, pvals, jaxpr, consts = linearize(traceable, *primals)
    107   else:
    108     out_primals, pvals, jaxpr, consts, aux = linearize(traceable, *primals, has_aux=True)

/usr/local/lib/python3.6/dist-packages/jax/interpreters/ad.py in linearize(traceable, *primals, **kwargs)
     93   _, in_tree = tree_flatten(((primals, primals), {}))
     94   jvpfun_flat, out_tree = flatten_fun(jvpfun, in_tree)
---> 95   jaxpr, out_pvals, consts = pe.trace_to_jaxpr(jvpfun_flat, in_pvals)
     96   out_primals_pvals, out_tangents_pvals = tree_unflatten(out_tree(), out_pvals)
     97   assert all(out_primal_pval.is_known() for out_primal_pval in out_primals_pvals)

/usr/local/lib/python3.6/dist-packages/jax/interpreters/partial_eval.py in trace_to_jaxpr(fun, pvals, instantiate, stage_out, bottom, trace_type)
    435   with new_master(trace_type, bottom=bottom) as master:
    436     fun = trace_to_subjaxpr(fun, master, instantiate)
--> 437     jaxpr, (out_pvals, consts, env) = fun.call_wrapped(pvals)
    438     assert not env
    439     del master

/usr/local/lib/python3.6/dist-packages/jax/linear_util.py in call_wrapped(self, *args, **kwargs)
    148     gen = None
    149 
--> 150     ans = self.f(*args, **dict(self.params, **kwargs))
    151     del args
    152     while stack:

/usr/local/lib/python3.6/dist-packages/jax_md/quantity.py in <lambda>(R, *args, **kwargs)
     29 def force(energy):
     30   """Computes the force as the negative gradient of an energy."""
---> 31   return grad(lambda R, *args, **kwargs: -energy(R, *args, **kwargs))
     32 
     33 

/usr/local/lib/python3.6/dist-packages/jax_md/smap.py in fn_mapped(R, species, species_count, **dynamic_kwargs)
    314       _kwargs = merge_dicts(kwargs, dynamic_kwargs)
    315       dr = _metric(R, R)
--> 316       for i in range(species_count):
    317         for j in range(species_count):
    318           s_kwargs = _kwargs_to_parameters((i, j), **_kwargs)

TypeError: 'JaxprTracer' object cannot be interpreted as an integer
```